### ISSUE TITLE
rasterio.block_windows, and overlapping computations in dask.array and xarray

### ISSUE THREAD

**Issue by ebo** [pangeo-data/pangeo#183]
I've just figured out part of my headaches -- in rasterio they are called block_windows, and in dask array (and xarray) they are called chunks...

Since I am working with geospatial data,working with rasterio is probably preferred, but I need to make this work.

From the rasterio concurent example <https://rasterio.readthedocs.io/en/latest/topics/concurrency.html> I see how to explicitly read/write ROI windows/chunks, but not how to ghost/overlap the data.  Can this be done in rasterio or do I have to use dask.array or xarray?  If I have to work with dask.array or xarray can you read/process the images in chunks?  As a note, when I am processing the images with the algorithms I am using I can ignore the geospatial information, and add it later.

Related to that, there are plenty of examples where you create delayed arrays and then process them in chunks, or read images in chunks, but immediately call .persist() -- which reads it all into memory.  I have yet figured out how to read, process and output images or even arrays in small chunks, and when I try to extend the examples in the way I need it always breaks.  

As a note, I have huge images (up to 3 giga-pixels times 4 to 8 bands).  I have to be able to read the image in chunks (with overlaps), process them, and save back out to a single image.  I've been poking at this as a side project for months, and now I am reading through the various documents trying to tease out how to make this work.  Any pointers, links, suggestions appreciated.

**Comment by ebo** [comment]
Another point.  How to you get the block address and start/stop x/y in dask.array and xarray?  I was planning to use the offset address or block index as an index offset to simplify the final integration step.

**Comment by rabernat** [comment]
I'm not sure I understand what this issue is about. Do you have a specific problem that you need help with? Or are you seeking general advice on architecture / design? Is this related to the use case you introduced in #153?

**Comment by ebo** [comment]
#153 was working through one of the tutorial examples that did not work out of the box, but I figured out what was wrong and how to work around it.  

I am working up to the application I mentioned above.  Basically I have a couple of algorithms I want to rewrite which would require processing overlapping (ghosted) chunks/windows.  The apps I am working on will only require a single pixel boundary.  That said I have to either access intermediate results or do a second pass across the chunk edges.

For clarity, is there a way to request an overlap or ghost with rasterio?  I have only seen it written about in dask.array or xarray.

**Comment by mrocklin** [comment]
My inclination here would be to load data normally and then use the [dask array map_overlap](http://dask.pydata.org/en/latest/array-api.html#dask.array.Array.map_overlap) method.  Does this suffice?

**Comment by ebo** [comment]
I will see if I can get that to work.  The problem I have mainly had before is that I have not been able to extend the example to work out of core with large imagery.  To stress test the examples I do things like set the memory limits to 3GB by "client = Client(memory_limit=3e9)", and then try to run a large image (300MP to 3GP multispectral imagery), and do anything with it.  While this might seem absurd, the smallest AWS allocations have something like 3GB of RAM, and my old code would process multiple 2.5GP images out of core and do it within the 3GB limits provided.  I would like to replace thousands of lines of my cobbled together code with dask et al.  That would make it more extensible and maintainable in the long term.

**Comment by mrocklin** [comment]
I see, so would it be correct to say that all we lack here is the ability to read large images in a chunk-wise fashion?  If so then presumably you could do something like the following:

```python
x = read_image_into_dask_array('myfile.fmt')
x.map_overlap(my_func, depth=1)
```

If so, then I'll ask the question back to you, do your images support being read in chunks?   If so, how would you get numpy array chunks out of them?

**Comment by ebo** [comment]
the old code supported reading in chunks (typically 1024x1024, but I tune this to keep everything within the footprint).  In that code I used GDAL's ReadAsArray(x, y,cols,rows).  Each image is single threaded, if if there is room for multiprocessing then I operated over different images -- basically embarrassingly parallel, and throw 15,000 to 20,000 of them at the cloud and let each one churn for 7 to 10 hours each.  I'm still amazed that you were able to figure out how to get GDAL to work in an distributed/parallel environment, and would love to get a delayed imread, or similar, to work.

How about this: lets see if we can come up with a simple example that applies a Gaussian blur to a large image with limited memory and small chunks with map_overlaps.   Getting that to work should direct me to where to study the examples.  It is very close to the example that is in overlap computing, but work start to finish.  I will be glad to help with testing and polishing an example.

**Comment by mrocklin** [comment]
It's probably not hard to do if you can recreate the GDAL workflow you have
in Python.  I recommend asking upstream at rasterio:
https://github.com/mapbox/rasterio/issues

**Comment by ebo** [comment]
My old workflow is in Python, but it is clunky as heck.  I'm hoping to replace that and go with something cleaner and more principled.  

While doing some background reading to prepare posting to rasterio, I came across the following issue <https://github.com/mapbox/rasterio/issues/1242#issuecomment-354562453> which probably will be a problem regardless.  

I'll give another attack using dask arrays and see where that takes me.  I may be able to do it all in DA and then try to update the profile later.  I am sure I will have more questions, but I will try to keep them to a minimum.

**Comment by mrocklin** [comment]
Yeah, you should definitely be using data sources that can be read in a
block-wise fashion.  If that's not the case then you probably won't be able
to do what you want to do in small memory, no matter what tools you use.

**Comment by ebo** [comment]
I was able to get the last image mosaics to process within a 3GB footprint, so the old dataset worked.  I will have to take a look at the most recent ones (which solve a lot of prior issues, and stack stack the different data layers).   So I am not sure, without digging around quite a bit, if all that data is tiled or if I stumbled upon a workaround by reading small windows.  But regardless, it is not given that the folks using these tools will have suitable data, and if not I should write a wrapper and issue a warning.

**Comment by mrocklin** [comment]
You can always read blocked patches out of non-blocked data, but it's very
very slow.

People who store their data in formats that don't support random blocked
access may just not be able to do good low-memory analysis.  I believe that
this is an issue that should be pushed upstream to data providers rather
than handled in computational libraries, where solutions aren't really
possible.

**Comment by ebo** [comment]
Agreed on pushing these issues upstream, and also that blocking non-blocked data can be slow, but if you can get it to work, it works.  

I just checked and fortunately all the different data streams we are currently working with are tiled (standardized 256x256 bloxks it appears).

**Comment by mrocklin** [comment]
Great, then all you need to do is to find a Python API that lets you get
numpy blocks out of those images.  I strongly recommend just asking a
question of rasterio, either on the issue tracker or stack overflow

**Comment by ebo** [comment]
I was hoping to figure out how to read in and process the data with dask arrays as well (not all of my data is geospatial).  What is the relationship between dask and rasterio?  I got the impression that they were the same folks.

As a note on stack overflow.  I do not know dask or rasterio well enough at this point to write up clear questions and solutions there.  I will post some more questions on rasterio's issue tracker.

BTW, I just got handed 4 new test images that are ~ 35000 x 225000 pixels (8GP with 5 bands of Uint16 data).  I have to figure this out.

**Comment by mrocklin** [comment]
The easiest and most common way to get data lazily into a dask array is
with an object that exposes the numpy slicing syntax.  See
http://dask.pydata.org/en/latest/array-creation.html#numpy-slicing

Failing that you would want a function that produced numpy array chunks,
and then you would use dask.delayed:
http://dask.pydata.org/en/latest/array-creation.html#using-dask-delayed

On Thu, Mar 29, 2018 at 11:06 AM, John (EBo) David <notifications@github.com

**Comment by digital-idiot** [comment]
Hello everyone, I'm also trying to achieve the same thing as @ebo, however I am not using `dask`. Is there any way `rasterio` can yield overlapped block with chosen depth or is there any plan to implement this? I also tried to do the same using `GDAL` directly, without any success. I have asked this on Stackoverflow([Stackoverflow Question](https://stackoverflow.com/questions/54501232/iteratively-load-image-block-by-block-where-blocks-are-partially-overlapped)) as well. IMHO this feature is extremely important for any kernel wise processing/filtering etc.

What are the current approaches (without manually windowing) to achieve the same? Working examples would be great.

**Comment by ebo** [comment]
Manually windowing is the only way I know of.  If you find another let 
me know.

Bet of luck!

**Comment by digital-idiot** [comment]
Hey @ebo,
[user2856@Stackoverflow](https://stackoverflow.com/users/737471/user2856) showed a nice way([Answer](https://stackoverflow.com/a/54525931/6422069)) of implementing overlapping window using `boundless` reading and also without it. Hope it helps.

**Comment by ebo** [comment]
Yes, I agree the stockoverflow post using rasterio was a nice 
presentation.  That said, my posting here was searching if we can do the 
same with the dask infrastructure.  If you look at the Overlapping 
Computations <http://docs.dask.org/en/latest/array-overlap.html> and 
Parallel Computing <http://xarray.pydata.org/en/stable/dask.html> 
documentation you will find that the chunks (or windows in rasterio 
parlance) are automatically broken up and processed.  That said, I have 
yet to find a way to tell me which row/column block is currently being 
processed.  For my case, I was interested in this because I can use it 
to partition a unique set of ID's that are guarantee to not be 
overlapping, and then post process collapse them to give a consecutive 
set of label ID's.  There are about 6 different ways to crack the nut 
but getting access to some of the chunks meta data proved to be 
difficult or not well documented.

Anyway, look at the overlapping and parallel coputations above.  They 
might give you what you need.

Hope this helps,

   EBo --

**Comment by philippjfr** [comment]
>That said, I have yet to find a way to tell me which row/column block is currently being processed.

This was also my stumbling block when I tried implementing a distributed regridding algorithm using dask. I can do the regridding on each ghosted chunk correctly but the exact shape of each chunk after regridding is dependent on the row/column it is in and I couldn't find a way to get access to that information.

**Comment by ebo** [comment]
On Feb 8 2019 6:30 AM, Philipp Rudiger wrote:
>>That said, I have yet to find a way to tell me which row/column block 
>> is currently being processed.
>
> This was also my stumbling block when I tried implementing a
> distributed regridding algorithm using dask. I can do the regridding
> on each ghosted chunk correctly but the exact shape of each chunk
> after regridding is dependent on the row/column it is in and I
> couldn't find a way to get access to that information.

Philipp,

It looks like you and I need to sync up off-list some time.  I cannot 
remember if I mentioned this in the previous posts, BUT in one 
implementation I used the property that if you enumerate each pixel in 
the image (through a generating function) and used that as the first ID 
/ offset in the image chunk for the cluster labeling, then you are 
guaranteed to have globally unique IDs after the first pass of the 
piece-wise labeling.  Then the union of the ghosted boundaries is 
trivial and simplifies the intermediate processing.

BTW, I posed a couple of branches to the dask-image repository which 
includes two versions of the watershed segmentation algorithm (the first 
follows dask-image conventions, and the second is a 
cython=>pure-python/numba implementation of the same.  I am trying to 
work my way up to a multi-scale OO segmentation algorithm which is an 
adaptation of stuff published decades ago, but only exists in expensive 
commercial apps.

   EBo --
